### PR TITLE
query: Replace dal with dal2, let's rock!

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Part of https://github.com/datafuselabs/databend/issues/3677

This PR intends to replace DAL with DAL2. In this PR, I expect to make `query` with `dal2` can **compile**, can **pass all tests**. All other issues will be addressed in the following PRs, like error handling and so on.

And you can see I add a lot of TODO's here, all of them will be resolved sooner or later. Also, please feel free to leave comments here, I will record them in the tracking issues so that others can involve in them too.

## Progress

All failed test cases will be listed here.

- [x] interpreters::interpreter_interceptor::test_interpreter_interceptor_for_insert (Fixed in https://github.com/datafuselabs/databend/pull/4081/commits/eb65db68583eb2b699d4dc0a58fa70118997ba3c)
- [x] metrics::test_metric_server (Fixed in https://github.com/datafuselabs/databend/pull/4081/commits/8c5046d4d9e29a373ba8af5ecfcca6c8dd722aca)
- [x] configs::test_initial_dir (Fixed in https://github.com/datafuselabs/databend/pull/4081/commits/ed0a2a5bccf4c6cf04a3187af2d3d493a8b9d8ee)
- [x] storages::fuse::operations::optimize::test_fuse_history_optimize
- [x] storages::fuse::operations::optimize::test_fuse_history_optimize_all
- [x] storages::fuse::operations::optimize::test_fuse_history_optimize_purge
- [x] storages::fuse::operations::purge_truncate::test_fuse_truncate_purge_stmt

## Changelog

- New Feature

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by BohuTANG** [comment]
@mergify update

File: query/src/sessions/query_ctx.rs [hunk 1]
```diff
@@ -295,37 +295,40 @@ impl QueryContext {
...
                 let conf = &storage_conf.s3;
-                Arc::new(S3::try_create(
-                    &conf.region,
-                    &conf.endpoint_url,
-                    &conf.bucket,
-                    &conf.access_key_id,
-                    &conf.secret_access_key,
-                    conf.enable_pod_iam_policy,
-                )?)
+                // TODO: Block on here to init the storage, we should find a better way.
```
Comment (BohuTANG) [inline query/src/sessions/query_ctx.rs:307]: Is it possible to change the function get_storage_accessor to async?
We have do a lot of work try to remove block_on in the past :)

Comment (Xuanwo) [reply query/src/sessions/query_ctx.rs:307]: Yes. It's the same concern @dantengsky raised too.

I plan to split `get_storage_accessor` into two functions (naming just for example):

- async `init_storage_accessor`: into storage accessor in the async way (and only once).
- sync `get_storage_accessor`: a simple function that returns an `Operator`, no extra APIs need to call.

**Comment by Xuanwo** [comment]
@dantengsky Hi, can you take time to help me investigate this test case: `storages::fuse::operations::optimize::test_fuse_history_optimize_purge`.

## Description

`test_fuse_history_optimize_purge` will fail for file not found. 

```txt
---- storages::fuse::operations::optimize::test_fuse_history_optimize_purge stdout ----
try to write _b/0b37e3fd476d4931ba3922fec3256def.parquet
finished write _b/0b37e3fd476d4931ba3922fec3256def.parquet
try to write _sg/44a70cabe43a497185245f78cc9a94ea
finished write _sg/44a70cabe43a497185245f78cc9a94ea
try to write _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- write here
finished write _ss/fcfc8115cdd843d1a2623ec0f9e217d8
try to write _b/a2e94e6b9c2f404bb9dcc40092bf3d10.parquet
finished write _b/a2e94e6b9c2f404bb9dcc40092bf3d10.parquet
try to write _sg/3116bcb3c8e24e4b9b9ba4def8c2448d
finished write _sg/3116bcb3c8e24e4b9b9ba4def8c2448d
try to stat _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- successful stat here
try to read _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- successful read here
try to write _ss/330b95dec66646f1afcd204f43b201f2
finished write _ss/330b95dec66646f1afcd204f43b201f2
try to stat _ss/330b95dec66646f1afcd204f43b201f2
try to read _ss/330b95dec66646f1afcd204f43b201f2
try to stat _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- successful stat here
try to read _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- successful read here
try to stat _sg/44a70cabe43a497185245f78cc9a94ea
try to read _sg/44a70cabe43a497185245f78cc9a94ea
try to stat _sg/3116bcb3c8e24e4b9b9ba4def8c2448d
try to read _sg/3116bcb3c8e24e4b9b9ba4def8c2448d
try to delete _b/0b37e3fd476d4931ba3922fec3256def.parquet
try to delete _sg/44a70cabe43a497185245f78cc9a94ea
try to delete _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- deleted here
try to stat _ss/330b95dec66646f1afcd204f43b201f2
try to read _ss/330b95dec66646f1afcd204f43b201f2
try to stat _ss/fcfc8115cdd843d1a2623ec0f9e217d8  <-- stat here
thread 'storages::fuse::operations::optimize::test_fuse_history_optimize_purge' panicked at 'we meet error here: "/tmp/.tmp7fpIc6/_ss/fcfc8115cdd843d1a2623ec0f9e217d8", No such file or directory (os error 2)', common/dal2/src/services/fs.rs:199:5
```

> In order to make the backtrace clear, I `panic` at `parse_io_error`.

It seems that we are trying to read the file after it has been deleted.

## Steps to reproduce

```shell
gh pr checkout 4081
RUST_BACKTRACE=1 cargo test -- "storages::fuse::operations::optimize::test_fuse_history_optimize_purge" --show-output
```

**Comment by Xuanwo** [comment]
All tests have been passed, let's get on the on the track to shining new world brought by DAL2.

**Review by PsiACE — APPROVED** [review]
:tada: Let's delete dal and rename dal2 to dal.

File: common/dal2/src/services/fs.rs [hunk 2]
```diff
@@ -104,10 +129,19 @@ impl Accessor for Backend {
             .await
             .map_err(|e| parse_io_error(&e, &path))?;
 
+        // `std::fs::File`'s errors detected on closing are ignored by
+        // the implementation of Drop.
+        // So we need to call `sync_all` to make sure all internal metadata
+        // have been flushed to fs successfully.
+        f.sync_all().await.map_err(|e| parse_io_error(&e, &path))?;
+
+        println!("finished write {}", args.path);
```
Comment (BohuTANG) [inline common/dal2/src/services/fs.rs:138]: I think this line no used anymore

File: common/dal2/src/services/fs.rs [hunk 3]
```diff
@@ -90,8 +100,23 @@ impl Accessor for Backend {
     }
 
     async fn write(&self, mut r: Reader, args: &OpWrite) -> Result<usize> {
+        println!("try to write {}", args.path);
```
Comment (BohuTANG) [inline common/dal2/src/services/fs.rs:103]: If we want to do print debug, it's better to use `tracing::debug!` instead of `println`.

Comment (Xuanwo) [reply common/dal2/src/services/fs.rs:103]: Yep, they should be revmoed.

File: common/dal2/src/services/fs.rs [hunk 4]
```diff
@@ -90,8 +100,23 @@ impl Accessor for Backend {
...
+
         let path = PathBuf::from(&self.root).join(&args.path);
 
+        // Create dir before write path.
+        //
+        // TODO(xuanwo): There are many works to do here:
+        //   - Is it safe to create dir concurrently?
+        //   - Do we need to extract this logic as new util functions?
+        //   - Is it better to check the parent dir exists before call mkdir?
+        let parent = path
```
Comment (BohuTANG) [inline common/dal2/src/services/fs.rs:113]: Does this only for `Fs` storage type?
We already inited the dir in:
https://github.com/datafuselabs/databend/blob/5bf157ed57e20566b99cc65ed296f2719175b121/query/src/bin/databend-query.rs#L49-L50

Comment (Xuanwo) [reply common/dal2/src/services/fs.rs:113]: I have removed `conf.initial_dir()` logic. As I think it's better to let DAL handle those work.

File: common/dal2/tests/it/services/s3.rs [hunk 5]
```diff
@@ -26,6 +27,34 @@ async fn builder() {
...
+    let mut r = op
+        .read("data/sample.csv")
+        .size(4 * 1024 * 1024)
+        .run()
         .await
         .unwrap();
+
+    let mut buf: Vec<u8> = Vec::new();
+    r.read_to_end(&mut buf).await.unwrap();
+    println!("{}", String::from_utf8(buf).unwrap());
```
Comment (BohuTANG) [inline common/dal2/tests/it/services/s3.rs:59]:
```suggestion

```

Comment (Xuanwo) [reply common/dal2/tests/it/services/s3.rs]: Those tests will be refactor into a behavior test for all DAL services.

File: query/tests/it/sessions/query_ctx.rs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_base::tokio;
+use common_exception::Result;
+use databend_query::configs::DiskStorageConfig;
+use databend_query::configs::S3StorageConfig;
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn test_get_storage_accessor_s3() -> Result<()> {
```
Comment (BohuTANG) [inline query/tests/it/sessions/query_ctx.rs:21]: :+1:

File: common/dal-context/src/context.rs [hunk 7]
```diff
@@ -110,26 +111,38 @@ impl DalContext {
...
 
 #[async_trait]
 impl Accessor for DalContext {
     async fn read(&self, args: &OpRead) -> DalResult<Reader> {
-        // TODO(xuanwo): Implement context callback reader to collect metrics.
-        self.inner.read(args).await
+        let metrics = self.metrics.clone();
+
+        // TODO(xuanwo): Maybe it's better to move into metrics.
+        self.inner.as_ref().unwrap().read(args).await.map(|reader| {
```
Comment (dantengsky) [inline common/dal-context/src/context.rs:127]: is it possible to eliminate this `unwrap` (and other `unwrap` in non-unit-testing codes)?

File: common/dal2/src/schema.rs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::str::FromStr;
+
+use super::error::Error;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Schema {
```
Comment (dantengsky) [inline common/dal2/src/schema.rs:19]: Is it a typo? maybe `Scheme` ?

File: common/dal2/tests/it/services/s3.rs [hunk 9]
```diff
@@ -26,6 +27,34 @@ async fn builder() {
         .credential(Credential::hmac("access-key", "secret-key"))
         .endpoint("http://localhost:9000")
         .finish()
+        .unwrap();
+}
+
+#[tokio::test]
+// TODO: we need to add an integration test for it.
```
Comment (BohuTANG) [inline common/dal2/tests/it/services/s3.rs:34]: Does this test will be covered by our statefultest?
https://github.com/datafuselabs/databend/blob/a479e82a5285645393eec05707ea2fe978b9e28e/.github/actions/test_stateful_standalone/action.yml#L52

File: common/dal2/src/services/fs.rs (hunk 4 above)
Comment (Xuanwo) [reply common/dal2/src/services/fs.rs:109]: Oh, I just found that this comment is on the `write` operation, please allow me to explain a bit more.

`Write` in DAL supports `write("a/b/c/d/e/f")` so the caller doesn't bother to call `mkdir` alike APIs. So we must create dir for `parent` here.

File: query/src/sessions/query_ctx.rs [hunk 10]
```diff
@@ -295,37 +294,36 @@ impl QueryContext {
...
+        let da: Arc<dyn Accessor> = match schema {
+            DalSchema::S3 => {
                 let conf = &storage_conf.s3;
-                Arc::new(S3::try_create(
-                    &conf.region,
-                    &conf.endpoint_url,
-                    &conf.bucket,
-                    &conf.access_key_id,
-                    &conf.secret_access_key,
-                    conf.enable_pod_iam_policy,
```
Comment (dantengsky) [inline query/src/sessions/query_ctx.rs:311]: @PsiACE  please take a look at this, not sure if we cound temporarily remove `pod_iam_policy`

File: common/dal2/tests/it/services/s3.rs (hunk 9 above)
Comment (Xuanwo) [reply common/dal2/tests/it/services/s3.rs:34]: > Maybe we should impl a mock to service the unit test?

I don't like the idea to mock storage services. In my experience, most of the mock behave inconsistently with the real one.

I prefer to test on a real storage service like local fs, minio.

File: common/dal-context/src/context.rs (hunk 7 above)
Comment (Xuanwo) [reply common/dal-context/src/context.rs:127]: There is no other way to access the `inner` outside `DalContext`, I believe it's safe here. But I do understand your concern, I will take it into consideration. How about resolving it in the next PRs?

File: common/dal2/src/schema.rs (hunk 8 above)
Comment (Xuanwo) [reply common/dal2/src/schema.rs:19]: Wow, english learnt.

File: common/dal-context/src/context.rs (hunk 7 above)
Comment (Xuanwo) [reply common/dal-context/src/context.rs:127]: Tracked in https://github.com/datafuselabs/databend/issues/3677

File: query/src/sessions/query_ctx.rs (hunk 10 above)
Comment (Xuanwo) [reply query/src/sessions/query_ctx.rs:311]: The aws-rust-sdk removed configs like `pod_iam_policy`. According to their README, it seems SDK will load those by default:

image

https://github.com/awslabs/aws-sdk-rust/

File: common/dal-context/src/context.rs (hunk 7 above)
Comment (dantengsky) [reply common/dal-context/src/context.rs:127]: > But I do understand your concern, I will take it into consideration.  

`unwrap` may panic. returns an Error might be better (at the Dal Layer)...

> How about resolving it in the next PRs?

Sure

File: query/src/sessions/query_ctx.rs (hunk 10 above)
Comment (Xuanwo) [reply query/src/sessions/query_ctx.rs:311]: I will add `Credential::Env` as default/fallback to load from env instead.

Comment (Xuanwo) [reply query/src/sessions/query_ctx.rs:311]: We will load credentials from env first and only set credentials if the user has input one. I think the current logic is the same as your said.

Maybe I need to call `aws_config::load_env()` here. But it's an async function, I will figure it out.

Comment (Xuanwo) [reply query/src/sessions/query_ctx.rs:311]: Resolved in https://github.com/datafuselabs/databend/pull/4081/commits/1bee64ba662bd5e4f04ff911c738a5ed2791933c

**Review by ZhiHanZ — APPROVED** [review]
awesome!

**Pull request merged**